src/core/regionmap.cpp

```diff
@@ -0,0 +1,316 @@
+            //  x                 y                 width             height            name
+                entry[0].toInt(), entry[1].toInt(), entry[2].toInt(), entry[3].toInt(), insertion
+            };
+        }
+    }
+    file.close();
+
+    project->mapSecToMapHoverName = qmap;
+
+    QFile binFile(region_map_layout_bin_path);
+    if (!binFile.open(QIODevice::ReadOnly)) return;
+    QByteArray mapBinData = binFile.readAll();
+    binFile.close();
+
+    for (int y = 0; y < layout_height_; y++) {
+        for (int x = 0; x < layout_width_; x++) {
+            int i = img_index_(x,y);
+            map_squares[i].secid = static_cast<uint8_t>(mapBinData.at(layout_index_(x,y)));
+            QString secname = (*(project->regionMapSections))[static_cast<uint8_t>(mapBinData.at(layout_index_(x,y)))];
+            if (secname != "MAPSEC_NONE") map_squares[i].has_map = true;
+            map_squares[i].mapsec = secname;
+            map_squares[i].map_name = sMapNamesMap.value(mapSecToMapEntry.value(secname).name);
+            map_squares[i].x = x;
+            map_squares[i].y = y;
+        }
+    }
+}
+
+void RegionMap::saveLayout() {
+    QString entries_text;
+    QString layout_text;
+
+    entries_text += "#ifndef GUARD_DATA_REGION_MAP_REGION_MAP_ENTRIES_H\n";
+    entries_text += "#define GUARD_DATA_REGION_MAP_REGION_MAP_ENTRIES_H\n\n";
+
+    for (auto sName : sMapNames) {
+        entries_text += "static const u8 sMapName_" + sName + "[] = _(\"" + sMapNamesMap.value(sName) + "\");\n";
+    }
+
+    entries_text += "\nconst struct RegionMapLocation gRegionMapEntries[] = {\n";
+
+    for (auto sec : *(project->regionMapSections)) {
+        if (!mapSecToMapEntry.contains(sec)) continue;
+        struct RegionMapEntry entry = mapSecToMapEntry.value(sec);
+        entries_text += "    [" + sec + "] = {" + QString::number(entry.x) + ", " + QString::number(entry.y) + ", " 
+            +  QString::number(entry.width) + ", " + QString::number(entry.height) + ", sMapName_" + entry.name + "},\n";
+    }
+    entries_text += "};\n\n#endif // GUARD_DATA_REGION_MAP_REGION_MAP_ENTRIES_H\n";
+
+    project->saveTextFile(region_map_entries_path, entries_text);
+
+    QByteArray data;
+    for (int m = 0; m < layout_height_; m++) {
+        for (int n = 0; n < layout_width_; n++) {
+            int i = img_index_(n,m);
+            data.append(map_squares[i].secid);
+        }
+    }
+    QFile bfile(region_map_layout_bin_path);
+    if (!bfile.open(QIODevice::WriteOnly)) return;
+    bfile.write(data);
+    bfile.close();
+}
+
+void RegionMap::saveOptions(int id, QString sec, QString name, int x, int y) {
+    resetSquare(id);
+    int index = getMapSquareIndex(x + this->padLeft, y + this->padTop);
+    if (!sec.isEmpty()) {
+        this->map_squares[index].has_map = sec == "MAPSEC_NONE" ? false : true;
+        this->map_squares[index].secid = static_cast<uint8_t>(project->regionMapSections->indexOf(sec));
```

Comment:
this requires map sections to be defined without any gaps in values, and without duplicates. do both of these requirements make sense? investigate